bftclient/src/bft_client.cpp

```diff
@@ -156,42 +204,88 @@ Reply Client::send(const MatchConfig& match_config,
     if (auto reply = wait()) {
       expected_commit_time_ms_.add(
           std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::steady_clock::now() - start).count());
+      reply_certificates_.clear();
       return reply.value();
     }
     metrics_.retransmissions.Get().Inc();
   }
 
   expected_commit_time_ms_.add(request_config.timeout.count());
-  outstanding_request_ = std::nullopt;
+  reply_certificates_.clear();
   throw TimeoutException(request_config.sequence_number, request_config.correlation_id);
 }
 
+std::deque<Reply> Client::sendBatch(std::deque<WriteRequest>& write_requests, const std::string& cid) {
+  std::deque<Reply> replies;
+  std::chrono::milliseconds max_time_to_wait = 0s;
+  MatchConfig match_config = writeConfigToMatchConfig(write_requests.front().config);
+  auto batch_msg = initBatch(write_requests, cid, max_time_to_wait);
+  auto start = std::chrono::steady_clock::now();
+  auto end = start + max_time_to_wait;
+  while (std::chrono::steady_clock::now() < end && replies.size() != pending_requests_.size()) {
+    bft::client::Msg msg(batch_msg);  // create copy here due to the loop
+    if (primary_) {
+      communication_->send(primary_.value().val, std::move(msg));
+    } else {
+      std::set<bft::communication::NodeNum> dests;
+      for (const auto& d : match_config.quorum.destinations) {
+        dests.emplace(d.val);
```

Comment:
Is it possible that the element already exists in dests and you just want to update it? If positive, use add_or_insert instead of emplace.

Reply:
The dests set created 2 rows before so it's supposed to be clean and the match config destinations should not hold duplicate destinations.